### PR TITLE
Shadowmonkers Grease Up

### PR DESCRIPTION
**Pull request by ayyaruq**
Speedmod is still being fickle and I'd like to later fix some stuff in Twin Snakes checklist, but tested that everything works. Commits broken up by the actual feature, and then some fixes to existing modules to replace dodgy workarounds with using the gauge. In one larger PR because I actually like ack yelling at me (and forgot to change branch).

### CODE REVIEW

File: src/parser/jobs/mnk/modules/Gauge.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
+import {t} from '@lingui/macro'
+
+import ACTIONS from 'data/ACTIONS'
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/Gauge.ts]: nit; no need to split deps

File: src/parser/jobs/mnk/modules/Gauge.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import {dependency} from 'parser/core/Module'
+import Combatants from 'parser/core/modules/Combatants'
+import {CounterGauge, Gauge as CoreGauge, TimerGauge} from 'parser/core/modules/Gauge'
+
+import DISPLAY_ORDER from './DISPLAY_ORDER'
+import {FORMS} from './Forms'
+
+// GL stack caps.
+export const MAX_STACKS = 3
+export const MAX_FASTER = 4
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/Gauge.ts]: ~~zoom zoom~~

File: src/parser/jobs/mnk/modules/Gauge.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+const GL_REFRESHERS: number[] = [
+	// kinda weird, only when going Coeurl->Opo-Opo and you need at least one stack
+	ACTIONS.FORM_SHIFT.id,
+	ACTIONS.SIX_SIDED_STAR.id,
+]
+
+// The duration for timeouts
+const GL_TIMEOUT_MILLIS = STATUSES.GREASED_LIGHTNING.duration * 1000
+
+export default class Gauge extends CoreGauge {
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/Gauge.ts]: nit; avoid `export default` in ts (and in general, but that can take it's time)

File: src/parser/jobs/mnk/modules/Gauge.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	static handle = 'greasedLightning'
+	static title = t('mnk.gauge.title')`Greased Lightning`
+	static displayOrder = DISPLAY_ORDER.GREASED_LIGHTNING
+
+	@dependency private combatants!: Combatants
+
+	private _stacks = this.add(new CounterGauge({
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/Gauge.ts]: nit; don't need to prefix privates in ts (technically)

File: src/parser/jobs/mnk/modules/Gauge.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	private _timer = this.add(new TimerGauge({
+		maximum: GL_TIMEOUT_MILLIS,
+		chart: {label: 'Duration', color: JOBS.PALADIN.colour},
+		onExpiration: () => this._stacks.reset(),
+	}))
+
+	protected init(): void {
+		super.init()
+
+		this.addHook('applybuff', {to: 'player', abilityId: STATUSES.FISTS_OF_WIND.id}, this.goFast)
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/Gauge.ts]: you get a  for all of these callback names tbqh

File: src/parser/jobs/mnk/modules/Fists.tsx [hunk 6]
```diff
@@ -159,7 +196,17 @@ export default class Fists extends Module {
 			value: this.getFistGCDCount(STATUSES.FISTS_OF_EARTH.id),
 		}))
 
-		// TODO: Add check for FoF GCD counts, per window, being larger than expected
+		this.suggestions.add(new TieredSuggestion({
+			icon: ACTIONS.FISTS_OF_WIND.icon,
+			content: <Trans id="mnk.fists.suggestions.fow.content">
+				Avoid swapping to <StatusLink {...STATUSES.FISTS_OF_WIND}/> while below {MAX_STACKS} stacks until you're about to execute a <StatusLink {...STATUSES.COEURL_FORM} /> skill. <StatusLink {...STATUSES.FISTS_OF_FIRE} /> offers more damage until you can get to GL{MAX_FASTER}.
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/Fists.tsx]: Should this read "...while below X stacks **unless** you're about to..."? The "until" reads weirdly in my head.

File: src/parser/jobs/mnk/modules/PerfectBalance.tsx [hunk 7]
```diff
@@ -1,25 +1,35 @@
...
+	@dependency private gauge!: Gauge
 
 	buffAction: Action = ACTIONS.PERFECT_BALANCE
 	buffStatus: Status = STATUSES.PERFECT_BALANCE
 
 	expectedGCDs = {
 		expectedPerWindow: 5,
 		suggestionContent: <Trans id="mnk.pb.suggestions.missedgcd.content">
-			Try to land 5 GCDs during every <ActionLink {...ACTIONS.PERFECT_BALANCE} /> window. If you cannot do this with full uptime and no clipping, consider adjusting your gearset for more Skill Speed.
+			Try to land at 5 GCDs in GL3, or 6 GCDs in GL4, during every <ActionLink {...ACTIONS.PERFECT_BALANCE} /> window. If you cannot do this with full uptime and no clipping, consider adjusting your gearset for more Skill Speed.
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/PerfectBalance.tsx]: Not sure about that "at" - intentional?

File: src/parser/core/modules/Gauge/CounterGauge.ts [hunk 8]
```diff
@@ -51,6 +52,12 @@ export class CounterGauge extends AbstractGauge {
 		this.chartOptions = opts.chart
 	}
 
+	getValueAt(timestamp: number) {
+		const counter = _.findLast(this.history, gauge => gauge.timestamp <= timestamp)
+		if (counter) { return counter.value }
```
[the remark this answers is not in the document]
Comment (ayyaruq) [reply src/parser/core/modules/Gauge/CounterGauge.ts]: I wanted to do that initially but afaik you mentioned it's only a feature in a newer version of Typescript. I might have misunderstood tho.

File: src/parser/jobs/mnk/modules/Gauge.ts (hunk 4 above)
Comment (ayyaruq) [reply src/parser/jobs/mnk/modules/Gauge.ts]: Is it not public by default? Idk if just me but a bunch of the standards we have for TS in xiva are the exact opposite to TC and that trips me up.

**Comment by ayyaruq** [comment]
Most feedback addressed, some stuff left for later since requires larger syntax changes with no job analysis effects and I'd rather do those in one big sweep.

**Pull request merged**